Retrieve macOS accent color

ebraminio opened the pull request:
This introduces getControlAccentColor and implements it in macOS.

Fixes material-foundation/flutter-packages#338

Guess some API reimagination would be needed but at least it works I can say,

<img width="803" alt="Screen Shot 2022-05-24 at 6 47 55 PM" src="https://user-images.githubusercontent.com/833473/170057895-6a663935-43b0-4505-b1c8-4855760505ea.png">

ebraminio committed:
Retrieve macOS accent color

This introduces getControlAccentColor and implements it in macOS.

guidezpl commented on lib/src/dynamic_color_builder.dart:
```diff
@@ -57,15 +62,28 @@ class DynamicColorBuilderState extends State<DynamicColorBuilder> {
-    });
+    if (corePalette != null) {
+      setState(() {
+        _light = corePalette?.toColorScheme();
```
```suggestion
        _light = corePalette.toColorScheme();
```

ebraminio replied:
This will say `The method 'toColorScheme' can't be unconditionally invoked because the receiver can be 'null'.`.

<img width="799" alt="image" src="https://user-images.githubusercontent.com/833473/170062734-fd3967d4-c522-4164-b196-2f4d6a2acbcc.png">

guidezpl replied:
Now that you're not checking whether it's null first, yeah.

guidezpl commented on lib/src/dynamic_color_plugin.dart:
```diff
@@ -14,6 +14,7 @@ class DynamicColorPlugin {
 
   /// The method name that the Kotlin plugin listens for.
   static const methodName = 'getCorePalette';
+  static const controlAccentColorMethodName = 'getControlAccentColor';
```
Needs documentation

guidezpl commented on lib/src/dynamic_color_builder.dart:
```diff
@@ -49,6 +49,11 @@ class DynamicColorBuilderState extends State<DynamicColorBuilder> {
+    // setState to update our non-existent appearance.
+    if (!mounted) return;
+
     CorePalette? corePalette;
```
I like the style you used better than what's here. Can you please rewrite `corePalette` fetching and `setState`-ing  in a single try-on block, like you did in lines 71-85?

ebraminio replied:
I think I addressed this hopefully.

ebraminio commented on example/pubspec.lock:
```diff
@@ -7,7 +7,7 @@ packages:
       name: async
```
Already done

ebraminio committed:
Partially address first round of reviews

ebraminio commented on macos/Classes/DynamicColorPlugin.swift:
```diff
@@ -0,0 +1,28 @@
+    switch call.method {
+    case "getControlAccentColor":
+      var nsColor = NSColor.systemBlue
+      if #available(macOS 10.14, *) {
```
Maybe return null if it Is below 10.14?

guidezpl replied:
That's consistent with the Android implementation

ebraminio replied:
Done

ebraminio commented:
Partially addressed some of the reviews hopefully, https://github.com/material-foundation/material-dynamic-color-flutter/pull/42/commits/14c65f5170843bc0ab9deec9eb7d39663e8c6ab5

ebraminio commented on lib/src/dynamic_color_plugin.dart:
```diff
@@ -28,4 +29,7 @@ class DynamicColorPlugin {
     return result == null ? null : CorePalette.fromList(result.toList());
   }
+
+  static Future<int> getControlAccentColor() async =>
```
I think I've addressed this hopefully

guidezpl replied:
Can you link to Apple docs which explain what it is? Also which versions of macOS have it, etc. What happens on other versions/platforms? Take a look at the documentation above for inspiration.

guidezpl replied:
https://developer.apple.com/design/human-interface-guidelines/macos/overview/whats-new-in-macos/#app-accent-colors

ebraminio commented on lib/src/dynamic_color_plugin.dart:
```diff
@@ -14,6 +14,7 @@ class DynamicColorPlugin {
 
   /// The method name that the Kotlin plugin listens for.
```
I think I've addressed this hopefully

ebraminio replied:
Done

ebraminio commented on macos/Classes/DynamicColorPlugin.swift:
```diff
@@ -0,0 +1,28 @@
+      var b: CGFloat = 0
+      var a: CGFloat = 0
+      nsColor.usingColorSpace(.sRGB)!.getRed(&r, green: &g, blue: &b, alpha: &a)
+      result(Int(a * 255) << 24 + Int(r * 255) << 16 + Int(g * 255) << 8 + Int(b * 255))
```
maybe better to float array here anyway but not sure if Windows counterpart maybe come later will use that or not. Some multiply it to `*255.99999` https://gist.github.com/gobijan/d724de27e2aff8131676 and some don't https://stackoverflow.com/questions/28644311/how-to-get-the-rgb-code-int-from-an-uicolor-in-swift

guidezpl replied:
Oh that's nasty, let's not do that

ebraminio replied:
OK

guidezpl commented on macos/Classes/DynamicColorPlugin.swift:
```diff
@@ -0,0 +1,28 @@
+      var g: CGFloat = 0
+      var b: CGFloat = 0
+      var a: CGFloat = 0
+      nsColor.usingColorSpace(.sRGB)!.getRed(&r, green: &g, blue: &b, alpha: &a)
```
Unfamiliar with switch but "unlikely" is not never so... :)

ebraminio replied:
Ok, considered its failure also

guidezpl commented on pubspec.yaml:
```diff
@@ -24,3 +24,6 @@ flutter:
         package: io.material.plugins.dynamic_color
         pluginClass: DynamicColorPlugin
+      macos:
+        package: io.material.plugins.dynamic_color
```
No idea, honestly

guidezpl committed:
Update README.md

ebraminio committed:
Address second round of reviews

ebraminio commented on macos/dynamic_color.podspec:
```diff
@@ -1,5 +1,10 @@
 Pod::Spec.new do |s|
   s.name             = 'dynamic_color'
+  s.authors          = 'Ebrahim Byagowi'
```
Should I put Google's name here or myself or both...?

guidezpl replied:
Material Flutter team + your name

ebraminio replied:
Done

ebraminio commented on test/dynamic_color_plugin_test.dart:
```diff
@@ -8,15 +10,35 @@ void main() {
   });
+
+  test('getControlAccentColor', () async {
+    const color = Color.fromARGB(12, 24, 123, 53);
```
Just a random value so we can know none of the channels have tampered.

guidezpl replied:
Nice

ebraminio commented:
> You'll also want to create a small example and tests so that the functionality you're adding can't be broken in the future.

Done the far I could.

guidezpl commented on lib/src/dynamic_color_builder.dart:
```diff
@@ -66,21 +66,25 @@ class DynamicColorBuilderState extends State<DynamicColorBuilder> {
-            brightness: Brightness.dark,
-          );
-        });
+        if (color == null) {
+          debugPrint('Control accent color result was null.');
+        } else {
```
I like this, please add this check above to avoid unecessary `setState`

ebraminio replied:
Done, also applied `corePalette?.toColorScheme` to `corePalette.toColorScheme`

guidezpl commented on lib/src/dynamic_color_plugin.dart:
```diff
@@ -32,7 +33,16 @@ class DynamicColorPlugin {
-  /// Returns the macOS' control accent color asynchronously in an [int] color.
-  static Future<int> getControlAccentColor() async =>
-      await channel.invokeMethod(controlAccentColorMethodName);
+  /// Returns the macOS' control accent color asynchronously in a [Color].
```
nit
```suggestion
  /// Returns the macOS' control accent color asynchronously as a [Color].
```

ebraminio replied:
Done

guidezpl commented on lib/src/dynamic_color_plugin.dart:
```diff
@@ -32,7 +33,16 @@ class DynamicColorPlugin {
-      await channel.invokeMethod(controlAccentColorMethodName);
+  /// Returns the macOS' control accent color asynchronously in a [Color].
+  ///
+  /// Supported since macOS 10.14, Mojave
```
```suggestion
  /// Supported since macOS 10.14 (Mojave).
  ///
  /// Completes will null on older versions of macOS 10.14, and on non-macOS platforms.
```

ebraminio replied:
Done

ebraminio committed:
Address third round of reviews

guidezpl committed:
improve consistency

guidezpl committed:
bump to 1.3.0

guidezpl committed:
tweak language

guidezpl committed:
bump to 1.3.0

guidezpl reviewed (approved):
```
  _    ___ _____ __  __ 
 | |  / __|_   _|  \/  |
 | |_| (_ | | | | |\/| |
 |____\___| |_| |_|  |_|
                        
```

I made small updates for consistency, and bumped the version to 1.3.0.

guidezpl committed:
formatting

guidezpl committed:
documentation sweep

guidezpl committed:
address lint

guidezpl committed:
Update pubspec.lock
